### PR TITLE
Add logging subsystem

### PR DESCRIPTION
**Pull request by collijk**
### Description

*Category*: feature
*Issues*: https://github.com/ihmeuw/vivarium/issues/282 and https://github.com/ihmeuw/vivarium/issues/283 

The primary goal of this PR is to add a logging subsystem into vivarium and to make it have as small an impact as 
possible on user facing code. 

**User-facing changes**
- `add_logging_sink`, `configure_logging_to_terminal`, and `configure_logging_to_file` have been moved from `vivarium.interface.utilities` to `vivarium.framework.logging.utilities`.  These functions are not used in `vivarium_cluster_tools` but may be alive in user code (Abie pointed out he uses it in https://github.com/ihmeuw/vivarium/issues/283 to suppress noisy logging, but part of the goal here is to give a more sensible way to do that).
- Output log format has changed.  Our old format is `{time} | {level} | {module}:{function}:{line} - {message}`.  There are two new formats.  The default (long) format is `{time} | {level} | {simulation_name}-{component_name}:{line} - {message}` if logging from a component that has requested a named logger and `{time} | {level} | {simulation_name}-{module}:{line} - {message}` if logging from a component that has requested an anonymous logger or if logging from the simulation context itself.  The idea here is that the component is going to be a more intuitive (and fine-grained) logging detail than the module for researchers and straightforward to translate to modules for engineers.  If that turns out not to be true, user space code can just request anonymous loggers.  Additionally, there is a short format log that does `{time} | {module}:{line} - {message}` that I think will be appropriate for things like running sims in parallel. Only the long format is currently exposed to users.
- The default logging level for both interactive and command line settings has been changed from `ERROR` at the command line and `DEBUG` in an interactive setting to  `INFO` in both settings (who thought that was a good idea originally :upside_down_face: ). 
- The logging level for the CLI is configurable from command line flags (`-v` logs at `DEBUG` level, `-q` logs at `WARNING` level)
- The logging level for the interactive sim is configurable from a `logging_verbosity` argument in the initializer (`2` logs at `DEBUG` level, `1` logs at `INFO` level [default], `0` logs at `WARNING` level). This argument is ignored if an external tool (like the command line interface) has configured logging.  External tools may want to log broader information to the same sinks and must have the ability to preempt.
- The following message levels have been changed:
  - Artifact manager messages are logged at `INFO` (formerly `DEBUG`).
  - The context message about unused configuration keys is now logged at `WARNING` (formerly `DEBUG`)
  - Results and performance metrics are now logged at `INFO` (formerly `DEBUG`)
- API docs have been setup in sphinx
- A concept stub for logging has been added to the `concept` section of the docs.
- User components can now request a logger from a new system of the builder `builder.logging.get_logger` and can, if desired, pass in the component name from which they want to log.

### Testing
CI and manual testing.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 import pytest
 import tables
+from _pytest.logging import LogCaptureFixture
```
Comment (rmudambi) [inline tests/conftest.py:5]: What is `_pytest`?

**Review by rmudambi — APPROVED** [review]

File: src/vivarium/framework/engine.py [hunk 2]
```diff
@@ -220,13 +228,13 @@ def report(self, print_results=True):
             self._population.get_population(True).index
         )
         if print_results:
-            logger.debug("\n" + pformat(metrics))
+            self._logger.info("\n" + pformat(metrics))
```
[the remark this answers is not in the document]
Comment (collijk) [reply src/vivarium/framework/engine.py:255]: I think this is a good point, but one to address in a different PR focused on the results system.

File: tests/conftest.py (hunk 1 above)
Comment (collijk) [reply tests/conftest.py:5]: Cargo culting.  This is a pattern from the loguru docs for making pytest work correctly with loguru.

File: src/vivarium/framework/logging/utilities.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    Parameters
+    ----------
+    output_directory
+        The directory to write the log file to.
+
+    """
+    log_file = output_directory / "simulation.log"
+    _add_logging_sink(
+        log_file,
+        verbosity=2,
```
[the remark this answers is not in the document]
Comment (collijk) [reply src/vivarium/framework/logging/utilities.py:51]: Will address in a separate PR

File: src/vivarium/framework/logging/utilities.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+        if self.long_format:
+            fmt += self.level + " | "
+
+        if self.long_format and "simulation" in record["extra"]:
+            if "component" in record["extra"]:
+                fmt += self.simulation_and_component + " - "
+            else:
+                fmt += self.simulation + " - "
+        else:
```
[the remark this answers is not in the document]
Comment (collijk) [reply src/vivarium/framework/logging/utilities.py:127]: For this if.

**Review by mattkappel — APPROVED** [review]

**Pull request merged**